torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -0,0 +1,277 @@
+        if self._in_pre_backward:
+            return output
+        if self._fsdp_param_group:
+            output = self._fsdp_param_group.post_forward(module, input, output)
+        self._training_state = TrainingState.IDLE
+        output = self._register_pre_backward_hook(output)
+        if self._mp_policy.output_dtype is not None:
+            with torch.profiler.record_function("FSDP::cast_forward_outputs"):
+                output = tree_map(
+                    functools.partial(
+                        _cast_floating_point_tensor, self._mp_policy.output_dtype
+                    ),
+                    output,
+                )
+        return output
+
+    def _pre_backward(self, *unused: Tuple[Any, ...]) -> None:
+        self._training_state = TrainingState.PRE_BACKWARD
+        if self._is_root and not self._root_post_backward_final_callback_queued:
+            log.info("root pre-backward hook")
+            self._register_root_post_backward_final_callback()
+        if self._fsdp_param_group:
+            self._fsdp_param_group._pre_backward(*unused)
+
+    def _root_post_backward_final_callback(self) -> None:
+        self._training_state = TrainingState.IDLE
+        if not self._is_root:
+            return
+        log.info("root post-backward final callback")
+        with torch.profiler.record_function("FSDP::root_post_backward_callback"):
+            for state_ref in self._all_state_refs:
+                state = state_ref()
+                assert state is not None, "FSDPState deallocated"
+                state._training_state = TrainingState.IDLE
+                if state._fsdp_param_group:
+                    state._fsdp_param_group.finalize_backward()
+            if not self._wait_for_grad_sync:
+                return
+            self._root_post_backward_final_callback_queued = False
+            self._post_forward_order.clear()
+
+    # Hook Registration #
+    def _register_pre_forward_hook(self) -> None:
+        if self._pre_forward_hook_handle is not None:
+            self._pre_forward_hook_handle.remove()
+        self._pre_forward_hook_handle = self._module.register_forward_pre_hook(
+            self._pre_forward, prepend=True, with_kwargs=True
+        )
+
+    def _register_post_forward_hook(self) -> None:
+        if self._post_forward_hook_handle is not None:
+            self._post_forward_hook_handle.remove()
+        self._post_forward_hook_handle = self._module.register_forward_hook(
+            self._post_forward, prepend=False
+        )
+
+    def _register_pre_backward_hook(self, output: Any) -> Any:
+        if not torch.is_grad_enabled():
+            return output
+
+        def _register_hook(tensor: torch.Tensor):
+            if tensor.requires_grad:
+                tensor.register_hook(self._pre_backward)
+                if self._fsdp_param_group:
+                    self._fsdp_param_group.needs_pre_backward_unshard = True
+            return tensor
```

Comment:
This "pre-backward" hook is logically like an `any()` backward hook over the module's forward output tensors.

For a module's forward output tensors, we only want the pre-backward hook to run once, and we want it to run upon the 1st output tensor's gradient being computed.

In this FSDP's design, we _do not_ use a post-accumulate grad hook for the complementary post-backward hook that frees parameters and reduce-scatter gradients. In contrast, we use a custom autograd function to insert a hook to run after the module's forward input tensors' gradients _and_ the module's parameters' gradients have been computed. This means that we will see one `(pre_backward, post_backward)` interval for every usage of the module in forward. In existing FSDP, we see only a _single_ `(pre_backward, post_backward)` interval that is the union of the intervals for this new FSDP.

The current approach uses a `bool` to implement the `any()` logic on this pre-backward hook. The post-backward hook makes sure to reset the `bool` in case of another distinct pre-backward later in backward.

I wonder if we can do something similar to the multi-grad hook and have this `any()` backward hook be a first-class citizen in core.